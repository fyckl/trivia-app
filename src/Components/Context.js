import React, { useContext, useState, useEffect } from 'react'

const Context = React.createContext()

function ContextProvider(props) {

    const [startQuiz, setStartQuiz] = useState(false)
    const [arrQuestions, setArrQuestions] = useState([])
    const [playAgain, setPlayAgain] = useState(false)
    const [correctAnswers, setCorrectAnswers] = useState(0)
    const [score, setScore] = useState(0)
    const [shuffleAnswers, setShuffleAnswers] = useState(false)

    function quizStart(){
      setShuffleAnswers(true)
      setStartQuiz(true)
    }

    function checkAnswers(){
      setShuffleAnswers(false)
      setPlayAgain(true)
    }

    useEffect(() => {
      setScore(prevState => prevState + correctAnswers)
    }, [correctAnswers])

    function reset(){
      setArrQuestions([])
      getQuestions()
      setPlayAgain(false)
      setCorrectAnswers(0)
      setShuffleAnswers(true)
    }

    function getQuestions(){
      fetch('https://opentdb.com/api.php?amount=5&difficulty=easy&type=multiple&encode=base64')
      .then(res=> res.json())
      .then(data=> setArrQuestions(data.results))
    }

    
    // console.log(arrQuestions)
    useEffect(() => {
      getQuestions()
    }, [])

    return(
        <Context.Provider 
            value={{
                quizStart, 
                startQuiz, 
                arrQuestions,
                setCorrectAnswers,
                checkAnswers,
                playAgain,
                correctAnswers,
                reset,
                score,
                shuffleAnswers
        }}>
            {props.children}
        </Context.Provider>
    )
}

export {ContextProvider, Context}