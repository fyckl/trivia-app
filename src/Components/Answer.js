import React, { useContext } from 'react'
import { Context, useAppContext } from "./Context"
import { Buffer } from 'buffer'

export default function Answer({ answers, correctAnswer}) {

  const {
    isPlayAgain, 
    selectedAnswers, 
    setSelectedAnswers
  } = useAppContext()
  
  const selectedAnswer = (answer) => answer.selectedAnswer !== answers
  const isSelected = selectedAnswers.every(selectedAnswer)

  const selectedAnswerClass = 
  !isPlayAgain ? 
  (isSelected ? "answer-button" : "selected-answer") : 
  (isSelected ? "answer-button" : "incorrect-answer")

  const buttonClass = 
  (isPlayAgain && (correctAnswer === answers)) ? 
  "correct-answer" : 
  selectedAnswerClass

  const decodedAnswers = Buffer.from(answers, 'base64').toString()
  function selectAnswer(){
    setSelectedAnswers(prevState => prevState.map(item => {
      return correctAnswer === item.id ? {...item, selectedAnswer: answers} : {...item}
    }))
  }

  return (
        <button 
          className={buttonClass} 
          onClick={selectAnswer} 
          disabled={isPlayAgain}
        >
          {decodedAnswers}
        </button>
  )
}