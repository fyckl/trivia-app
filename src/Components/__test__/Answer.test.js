import React from "react"
import Answer from "../Answer"
import * as Context  from "../Context"
import "@testing-library/jest-dom/extend-expect"
import { shallow, configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

test("answer renders correctly and matches snapshot", () => {
    const contextValues = {
        isPlayAgain: false,
        selectedAnswers:[],
        setSelectedAnswers:jest.fn()
    }
    jest
      .spyOn(Context, 'useAppContext')
      .mockImplementation(() => contextValues);

    const wrapper = shallow(<Answer answers={[]} />)

    expect(wrapper.html()).toMatchSnapshot()
})