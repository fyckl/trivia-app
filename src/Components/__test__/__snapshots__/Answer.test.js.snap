// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`answer renders correctly and matches snapshot 1`] = `"<button class=\\"answer-button\\"></button>"`;
